### ISSUE TITLE
sym.S.ImaginaryUnit == 1j  gives False

### ISSUE THREAD

**Issue by mohamed-180** [sympy/sympy#26996]
Based on 

> Zero is a singleton, and can be accessed by ``S.Zero``

And 

```
In [1]: sym.S.Zero == 0                                                                       
Out[1]: True              
```

Why this gives false

```
In [219]: sym.I == 1.0j                                                                         
Out[219]: False                                                                                 

In [220]: sym.S.ImaginaryUnit == 1j                                                             
Out[220]: False                            
```

**Comment by oscarbenjamin** [comment]
Note also in SymPy 1.13:
```python
In [1]: S(1) == S(1.0)
Out[1]: False
```
In other words floats do not compare equal to Integer/Rational in SymPy because `==` is for structural equality. See gh-26817.

I am wondering though whether we should relax this specifically for comparison with non-SymPy types though...

**Comment by oscarbenjamin** [comment]
This one should give True:
```python
In [5]: 1.0*I == 1j
Out[5]: False
```

**Comment by oscarbenjamin** [comment]
I'm not sure what happens here:
```python
In [2]: sympify(1j) == 1.0*I
Out[2]: True

In [3]: 1j == 1.0*I
Out[3]: False

In [4]: 1.0*I == sympify(1j)
Out[4]: True

In [5]: 1.0*I == 1j
Out[5]: False
```

**Comment by oscarbenjamin** [comment]
The fix:
```diff
diff --git a/sympy/core/basic.py b/sympy/core/basic.py
index caab088cf2..92c2012968 100644
--- a/sympy/core/basic.py
+++ b/sympy/core/basic.py
@@ -458,12 +458,25 @@ def _do_eq_sympify(self, other):
         that after conversion, `==`  is used again since it is not
         necessarily clear whether `self` or `other`'s __eq__ method needs
         to be used."""
-        for superclass in type(other).__mro__:
-            conv = _external_converter.get(superclass)
-            if conv is not None:
-                return self == conv(other)
+
+        # We need a special case for complex here because it does not
+        # correspond directly to any particular sympy type whose __eq__ method
+        # could handle it.
+        if isinstance(other, complex):
+            return self == _sympify(other)
+
+        # If downstream code has defined sympifiable types then we just have
+        # to sympify them here.
+
+        if _external_converter:
+            for superclass in type(other).__mro__:
+                conv = _external_converter.get(superclass)
+                if conv is not None:
+                    return self == conv(other)
+
         if hasattr(other, '_sympy_'):
             return self == other._sympy_()
+
         return NotImplemented
 
     def __eq__(self, other):
```

**Comment by sylee957** [comment]
`S.Zero` and `0` are also equal by hash, but `S.ImaginaryUnit` and `1.0j` are not

```
>>> from sympy import *
>>> S.Zero == 0
True
>>> S.One == 1.0
False
>>> hash(S.Zero) == hash(0)
True
>>> hash(S.ImaginaryUnit)
-1551392344412138492
>>> hash(1.0j)
1000003
```

I may need to revisit how `hash` is supposed to work, 
however, that was obviously outside the discussion of author (who only understood via singleton numbers) and could be overlooked.

However, if `eq` should imply `hash` by principle, I think that we can't do this because it breaks the principle.
https://www.asmeurer.com/blog/posts/what-happens-when-you-mess-with-hashing-in-python/

**Comment by oscarbenjamin** [comment]
We would need these two to be equal:
```python
In [1]: hash(1j)
Out[1]: 1000003

In [2]: hash(1.0*I)
Out[2]: 2798142013897519853
```

**Comment by mohamed-180** [comment]
> No, the article says the `==` equality should imply hash equality, so if `==` is `False`, then hash being either `True` or `False` is sound behavior Also, that is pretty much how hash conflict works.
> 
> However, only the cases where `==` returns `True` but `hash` being unequal can cause problems.

I think `==` does not imply hash equality, see :

```
class Deff_Hashs:
    i = 0
    def __init__(self, x) -> None:
        self.i = __class__.i
        self.x = x
        __class__.i += 1

    def __hash__(self) -> int:
        return hash(self.i)

    def __eq__(self, value) -> bool:
        return self.x == value.x

h1 = HashOne(12) 
h2 = HashOne(12) 
```
Here  `h1` and `h2` are equal by `__eq__` function but have unequal hashs

**Comment by oscarbenjamin** [comment]
> Here `h1` and `h2` are equal by `__eq__` function but have unequal hashs

If `__eq__` return True but the hashes are unequal then that is a bug and either the `__eq__` or the `__hash__` method should be changed.

**Comment by mohamed-180** [comment]
I know that was a bad decision, But i was think that `==` implies equality of hash by default.

**Comment by sylee957** [comment]
Obviously, it is still possible to define `__eq__` or `__hash__` in some nonsense,
However, the point is that defining `__eq__` and `__hash__`  should be done very carefully
If Python stdlib or some other open source libraries find such issues, you can easily see that they identify them as bug.

**Comment by oscarbenjamin** [comment]
The awkwardness here comes from the fact that SymPy does not have a `Complex` expression type (unlike e.g. SymEngine) that corresponds to the python `complex` type.

I think we probably want `__eq__` and `__hash__` to emulate the effect of having a `Complex` type. At least if `__eq__` does that then `__hash__` needs to match it. I'm not sure how `complex.__hash__` works in general:
```python
In [11]: hash(5+7j)
Out[11]: 7000026

In [12]: 1000003*7 + 5
Out[12]: 7000026

In [13]: hash(.5+.7j)
Out[13]: 5995191823853203968

In [14]: 1000003*.7 + .5
Out[14]: 700002.6

In [15]: 1000003*hash(.7) + hash(.5)
Out[15]: 1614096101641409619603968

In [16]: (1000003*hash(.7) + hash(.5)) % 2**64
Out[16]: 5995191823853203968
```
We have `pure_complex` that could be used for this.

There are other examples though:
```python
In [20]: hash((1, 2))
Out[20]: -3550055125485641917

In [21]: hash(Tuple(1, 2))
Out[21]: -3550055125485641917

In [27]: hash(frozenset([1, 2]))
Out[27]: -1826646154956904602

In [28]: hash(FiniteSet(*[1, 2]))
Out[28]: -2326809509863296318

In [29]: FiniteSet(*[1, 2]) == frozenset([1, 2])
Out[29]: True
```

**Comment by sylee957** [comment]
CPython may be using this implementation
https://github.com/python/cpython/blob/6754566a51a5706e8c9da0094b892113311ba20c/Objects/complexobject.c#L452-L472
So `(1000003*hash(.7) + hash(.5)) % 2**64` may be similar to this implementation, but I'm not 100% sure that I investigate such low level details.

Again representing complex number as add/mul that SymPy had taken can be challenging,
and I think that it gets into more wider scope to change `Add.__hash__` and `Mul.__hash__` as well

**Comment by sylee957** [comment]
However, my suggestion to change hash for Add, Mul looks like:
```
def __hash__(self):
    return reduce(lambda x, y: (x + hash(y)) % 2**64, self.args, 0)

def __hash__(self):
    return reduce(lambda x, y: (x * hash(y)) % 2**64, self.args, 0)
```
I want to avoid going more deep into complex number decomposition, because I we get much more complexity into handling things like assumptions

**Comment by oscarbenjamin** [comment]
> However, my suggestion to change hash for Add, Mul looks like:

SymPy depends quite significantly on having an efficient `__hash__` implementation as is used by hashable content etc. This is needed because `__hash__` is used in every operation that is cached and most basic things are cached.

I would expect that your suggestion would have a noticeable performance impact for large expressions.

> decomposing complex number to real and imaginary part, which is often confusing as we have symbolic expressions with complex assumptions, or such.

We don't need to worry about assumptions, symbols or more complex expressions or anything. The only possibilities here are of these structural forms:
```python
Add(Float(r), Mul(Float(i), ImaginaryUnit))
Mul(Float(i), ImaginaryUnit)
```
If we wanted Float and Rational to compare equal then we would also need to handle Rational in place of Float as well as the additional form:
```python
Add(Float(r), ImaginaryUnit)
```

**Comment by sylee957** [comment]
That's another way around, but it just looks like it involves adding some front-faced if statement to `Add` or `Mul`, that's also something I find very trivial and unnecessary feature to introduce at all.

It is much more easier to resolve this issue without planned.

Maybe the issue with FiniteSet and frozenset should be continued in another issue. I also suggest to simply define equality between `FiniteSet` and `frozenset` to be false, because they come from different structure/class, if there is no obvious solution.
(The case with `tuple` is different because Python allows efficient and semantic computing of hash for tuples)

**Comment by oscarbenjamin** [comment]
> It is much more easier to resolve this issue without planned.

The original issue is about `__eq__` rather than `__hash__` which probably should be fixed and I have shown the fix above.

**Comment by oscarbenjamin** [comment]
> Maybe I misunderstood something, but there was a PR to change Float and Rational to be unequal,
> and I just understood that as a motivation to treat == between objects more strict.

The motivation from my perspective was to make it so that `==` between SymPy expressions handles structural equality consistently. The reason for this is more for internal consistency rather than what I want from a user interface. It is not simply about being "more" strict but making it so that `a == b` is usable for internal code that manipulates expressions like:
https://github.com/sympy/sympy/blob/adff08d8fa1be5623017f1d46c239392d8c81399/sympy/core/basic.py#L1728-L1733 

One difficulty with SymPy's Basic design is that there is no distinction between the internal and external interface. Our internal code has to use the same `==` and `!=` operators that are used by users and in turn dicts and sets use those same operators as well even if we choose not to use them in the codebase directly. Since basically everything in SymPy operates under a dict-based cache system everything is always affected by the way that dicts handle expressions:
```python
In [15]: d = {1: 'int', 1.0: 'float'}

In [16]: d[1]
Out[16]: 'float'
```
The only reason this didn't cause massive problems all the time was because `float.__hash__` was broken.

If an expression had an internal expression of a different type like `expr.rep` then we could define `__eq__` differently for the internal type rather than for the user-facing type. Probably then we could say that `__eq__` should be for semantic equality rather than structural equality which is possibly a better default for users. Then users could do e.g. `solve(x**2 == 2)`. Maybe then we could use the consistent `__hash__/__eq__` methods of the internal type to make it so that all expressions are uniquely represented in memory and then `a is b` could be for structural equality.

It might be awkward to use such expressions in dicts and sets or other situations but SymPy's internal code could work with the internal type that would use structural equality and would play nice with all Python's basic data structures. You could even define `__le__` etc to do structural comparisons so that `sorted(expressions)` works like `list(ordered(expressions))` when working with the internal expression type.

I don't think that I would have actually merged the change to Float/Rational comparisons even though I did the initial work for it. That is why I tried to extract what I thought were useful pieces from gh-20033 rather than asking for it to be merged. There should have been some objection rather than it being self merged like this since it is a significant breaking change and iss obviously controversial (obvious meaning that you could know it would be controversial without needing to wait for anyone else to give feedback):
https://github.com/sympy/sympy/pull/25614#issuecomment-1714546309.

In any case the issue from my perspective was not about SymPy expressions comparing equal to other objects as such. Just there is also a kind of invariant that if `a` is a sympy expression and `b` is some other kind of object then the following is usually expected to hold:
```python
(a == b) == (a == _sympify(b))
```
In other words SymPy objects should be usable interchangeably with their unsympified counterparts when possible and for that to be true you want `==` to be consistent before and after sympification of the other object.

I could certainly be persuaded that we should revert the main Float/Rational change as it seems to cause trouble for users and downstream (see the back-references under gh-25614).

**Comment by sylee957** [comment]
> In any case the issue from my perspective was not about SymPy expressions comparing equal to other objects as such. Just there is also a kind of invariant that if a is a sympy expression and b is some other kind of object then the following is usually expected to hold:
> (a == b) == (a == _sympify(b))
> In other words SymPy objects should be usable interchangeably with their unsympified counterparts when possible and for that to be true you want == to be consistent before and after sympification of the other object.

I think that changing the behavior of `__eq__` from `False` to `True` may have some impact for backward compatibility. 
Users can also find the changes of truth value like this difficult to rely on to use to be useful, at least for certain period of time.
So we may need to be clear about getting ideas aligned and move carefully

**Comment by smichr** [comment]
Python doesn't have a native way to represent an exact imaginary unit. SymPy does. I don't think we should think of how to make this true any more than trying to make `pi == math.pi`. I would recommend closing this issue and reintroducing any ideas of merit as a new issue.
